### PR TITLE
fix(instrumenter): Handle TypeScript NonNullExpression

### LINKED ISSUE #3080: Optional chaining mutator fails when followed by TypeScript `!`
**Summary**

The optional chaining mutator creates different (failing) original code when followed by a TypeScript `!`.

Repro repo here: https://github.com/hugo-vrijswijk/stryker-repro

The following code:

```typescript
export function foo(input: { id: number | undefined } | undefined) {
  return input?.id!.toString();
}
```

The function body is mutated to:

```typescript
return (stryMutAct_9fa48("1") ? input.id : (stryCov_9fa48("1"), input?.id))!.toString();
```

Removing the `!`:

```typescript
export function foo(input: { id: number  } | undefined) {
  return input?.id.toString();
}
```

Mutates the function body to:

```typescript
return stryMutAct_9fa48("1") ? input.id.toString() : (stryCov_9fa48("1"), input?.id.toString());
```

Which behaves differently, even though the `!` should not make a difference. In the repro repo, this causes the initial test run to fail. Even though running `npm test` works fine

**Stryker config**

```json
{
  "$schema": "./node_modules/@stryker-mutator/core/schema/stryker-schema.json",
  "packageManager": "npm",
  "reporters": ["html", "clear-text", "progress"],
  "testRunner": "mocha",
  "coverageAnalysis": "perTest",
  "buildCommand": "tsc -b",
  "mochaOptions": {
    "spec": ["dist/**/*.test.js"]
  }
}

```

**Test runner config**

None, other than pointing Mocha to compiled code (see mochaOptions)

**Stryker environment**

```
├─┬ @stryker-mutator/core@5.2.1
│ ├─┬ @stryker-mutator/api@5.2.1
│ ├─┬ @stryker-mutator/instrumenter@5.2.1
│ │ ├── @stryker-mutator/api@5.2.1 deduped
│ │ ├── @stryker-mutator/util@5.2.1 deduped
│ ├─┬ @stryker-mutator/util@5.2.1
├─┬ @stryker-mutator/mocha-runner@5.2.1
│ ├── @stryker-mutator/api@5.2.1 deduped
│ ├── @stryker-mutator/util@5.2.1 deduped
```

This happens in Stryker 5.2.1 and above.

```
"mocha": "^9.1.0",
"typescript": "^4.3.5"
```

**Test runner environment**

```shell
mocha dist/**/*.test.js
```

**Your Environment**

| software         | version(s)
| ---------------- | -------
| node             | 12
| npm              | 6
| Operating System | Windows (WSL)

**Add stryker.log**

### PR DESCRIPTION
**Pull request by simondel**
Fixes #3080

### CODE REVIEW

File: packages/instrumenter/src/mutant-placers/expression-mutant-placer.ts [hunk 1]
```diff
@@ -60,7 +60,7 @@ function isMemberOrCallExpression(path: NodePath) {
 }
 
 function isMemberExpression(path: NodePath): path is NodePath<types.MemberExpression | types.OptionalMemberExpression> {
-  return path.isMemberExpression() || path.isOptionalMemberExpression();
+  return path.isMemberExpression() || path.isOptionalMemberExpression() || path.isTSNonNullExpression();
```
Comment (nicojs) [inline packages/instrumenter/src/mutant-placers/expression-mutant-placer.ts:63]: This breaks typesafety, as the return type specifies it to be a member expression. Also, the method name is no longer correct 🤷‍♂️

File: packages/instrumenter/test/unit/mutant-placers/expression-mutant-placer.spec.ts [hunk 2]
```diff
@@ -75,10 +75,11 @@ describe(expressionMutantPlacer.name, () => {
         ['foo?.bar()', (p) => p.isOptionalMemberExpression() && types.isIdentifier(p.node.property, { name: 'bar' })],
         ['foo.bar?.baz', (p) => p.isMemberExpression() && types.isIdentifier(p.node.property, { name: 'bar' })],
         ['foo?.bar.baz', (p) => p.isOptionalMemberExpression() && types.isIdentifier(p.node.property, { name: 'bar' })],
+        ['foo?.bar!.baz', (p) => p.isTSNonNullExpression()],
       ];
       falsePointers.forEach(([js, query]) => {
         it(`should not allow placing in \`bar\` of \`${js}\``, () => {
-          const path = findNodePath(parseJS(js), query);
+          const path = findNodePath(parseTS(js), query);
           expect(expressionMutantPlacer.canPlace(path)).false;
```
Comment (nicojs) [inline packages/instrumenter/test/unit/mutant-placers/expression-mutant-placer.spec.ts:83]: Nice addition! Could we also add a bon-null assertion to the integration test and update the snapshot?

**Review by nicojs — CHANGES_REQUESTED** [review]
Quick review from my phone 🤳.

Great to see it's only a small change

File: packages/instrumenter/src/mutant-placers/expression-mutant-placer.ts (hunk 1 above)
Comment (simondel) [reply packages/instrumenter/src/mutant-placers/expression-mutant-placer.ts:63]: Good point! I forgot to change that

Comment (simondel) [reply packages/instrumenter/src/mutant-placers/expression-mutant-placer.ts:63]: The NonNullExpression seems like a MemberExpression to me, so I think the method name does seem valid.

File: packages/instrumenter/test/unit/mutant-placers/expression-mutant-placer.spec.ts (hunk 2 above)
Comment (nicojs) [reply packages/instrumenter/test/unit/mutant-placers/expression-mutant-placer.spec.ts:83]: Did that ✅

**Pull request merged**